src/Java.Interop.BootstrapTasks/Java.Interop.BootstrapTasks/JdkInfo.cs

```diff
@@ -14,74 +17,265 @@ public class JdkInfo : Task
+
+		[Required]
+		public  ITaskItem       PropertyFile        { get; set; }
+
 		[Required]
-		public ITaskItem Output { get; set; }
+		public  ITaskItem       MakeFragmentFile    { get; set; }
 
 		[Output]
-		public string JavaSdkDirectory { get; set; }
+		public  string          JavaHomePath        { get; set; }
+
+		static  readonly  bool  OnMacOS = Directory.Exists ("/Applications");
 
 		public override bool Execute ()
 		{
 			Log.LogMessage (MessageImportance.Low, $"Task {nameof (JdkInfo)}");
-			Log.LogMessage (MessageImportance.Low, $"  {nameof (Output)}: {Output}");
+			Log.LogMessage (MessageImportance.Low, $"  {nameof (JdksRoot)}: {JdksRoot}");
+			Log.LogMessage (MessageImportance.Low, $"  {nameof (MakeFragmentFile)}: {MakeFragmentFile}");
+			Log.LogMessage (MessageImportance.Low, $"  {nameof (MaximumJdkVersion)}: {MaximumJdkVersion}");
+			Log.LogMessage (MessageImportance.Low, $"  {nameof (PropertyFile)}: {PropertyFile}");
+
+			var maxVersion  = GetMaxJdkVersion ();
+			var java_home   = GetJavaHomePathFromEnvironment ();
+			if (java_home != null) {
+				var java_home_v = GetVersionFromPath (java_home);
+				if (maxVersion != null && java_home_v != null && java_home_v > maxVersion) {
+					Log.LogMessage (MessageImportance.Low, $"  Skipping JAVA_HOME default value of `{java_home}` as it exceeds MaximumJdkVersion={MaximumJdkVersion}.");
+					java_home = null;
+				}
+				if (java_home != null && !Directory.Exists (Path.Combine (java_home, "include"))) {
+					Log.LogMessage (MessageImportance.Low, $"  Skipping JAVA_HOME default value of `{java_home}` as it does not contain an `include` subdirectory.");
+					java_home = null;
+				}
+			}
+			java_home = java_home ?? GetJavaHomePathFromMachine (maxVersion);
 
-			var javaSdkPath = GetJavaSdkPath ();
-			if (string.IsNullOrEmpty(javaSdkPath)) {
-				Log.LogError ("JavaSdkPath is blank");
+			if (string.IsNullOrEmpty (java_home)) {
+				Log.LogError ("Could not determine JAVA_HOME location. Please set JdksRoot or export the JAVA_HOME environment variable.");
 				return false;
 			}
 
-			Log.LogMessage (MessageImportance.Low, $"  JavaSdkPath: {javaSdkPath}");
+			var jarPath     = "";
+			var javacPath   = "";
+			var jdkJvmPath  = "";
+			var includes    = new List<string> ();
+
+			if (Path.DirectorySeparatorChar == '\\') {
+				jarPath     = Path.Combine (java_home, "bin", "jar.exe");
+				javacPath   = Path.Combine (java_home, "bin", "javac.exe");
+				jdkJvmPath  = Path.Combine (java_home, "jre", "bin", "server", "jvm.dll");
+			} else {
+				jarPath     = Path.Combine (java_home, "bin", "jar");
+				javacPath   = Path.Combine (java_home, "bin", "javac");
+				if (OnMacOS) {
+					jdkJvmPath  = Directory.EnumerateFiles (java_home, "libjli.dylib", SearchOption.AllDirectories)
+						.FirstOrDefault ();
+				} else {
+					jdkJvmPath  = Path.Combine (java_home, "jre", "bin", "server", "libjvm.so");
+				}
+			}
```

Comment:
Wouldn't it be better to determine the file extension somewhere at the start of the task and remove need for this `if` statement? The difference between macOS and Linux library names would need to stay, of course, but the "complexity" (and repetitiveness) of code would be reduced

Reply:
This single filename has to be special-cased, yes, but the others don't need to be.